### PR TITLE
Expanded BLE Advertisement Properties

### PR DESCRIPTION
**Pull request by davidtaylor-juul**
Expands the Advertisement class to include the majority of the information that can be extracted from BLE advertisement packets.

### CODE REVIEW

File: core/src/jsMain/kotlin/Peripheral.kt [hunk 1]
```diff
@@ -205,11 +205,13 @@ public class JsPeripheral internal constructor(
...
-            .characteristics
+            .firstOrNull { it.serviceUuid == characteristic.serviceUuid }
+            ?.characteristics
+            ?: throw IOException("Service ${characteristic.serviceUuid} not found")
         return characteristics
-            .first { it.characteristicUuid == characteristic.characteristicUuid }
-            .bluetoothRemoteGATTCharacteristic
+            .firstOrNull { it.characteristicUuid == characteristic.characteristicUuid }
+            ?.bluetoothRemoteGATTCharacteristic
+            ?: throw IOException("Characteristic ${characteristic.characteristicUuid} not found")
```
Comment (davidtaylor-juul) [inline core/src/jsMain/kotlin/Peripheral.kt:214]: This is unrelated - was added when debugging, can remove if not wanted

Comment (twyatt) [reply core/src/jsMain/kotlin/Peripheral.kt:214]: Ya, please revert. We'll make this change (or similar) when we close #38.

**Comment by twyatt** [comment]
Looking at this implemented, having to provide a manufacturer ID is quite awkward for a library user. Read through the BLE spec a bit and couldn't find much in the way that hints why Android and JavaScript went that route for their APIs.

How ugly is the JavaScript code for getting the bytes (without having to provide the manufacturer ID)?

File: core/src/jsMain/kotlin/external/BluetoothAdvertisingEvent.kt [hunk 2]
```diff
@@ -6,6 +6,11 @@ import org.w3c.dom.events.Event
  * https://webbluetoothcg.github.io/web-bluetooth/#bluetoothadvertisingevent
  */
 internal abstract external class BluetoothAdvertisingEvent : Event {
-    val rssi: Int
     val device: BluetoothDevice
+    val uuids: Array<String>
+    val name: String?
+    val rssi: Int?
+    val txPower: Int?
+    val manufacturerData: Any?
```
Comment (twyatt) [inline core/src/jsMain/kotlin/external/BluetoothAdvertisingEvent.kt:14]: The docs say this is map-like. Does that mean that we could just mark it as a `Map` in Kotlin? Since I assume the functions and properties would align with that of a typical map (assuming Kotlin and JavaScript have similar naming for their `Map` properties/functions)?

Wondering if we have access to `keys` and could ask for element at index `0` or use `first`, to be able to ultimately grab the manufacturer ID.

Comment (davidtaylor-juul) [reply core/src/jsMain/kotlin/external/BluetoothAdvertisingEvent.kt:14]: The docs indicate they conform to `maplike` but I've found this to be only partially true for the current (experimental) implementation of both the service and manufacturer data. Typecasting to a `Map` in Kotlin fails as none of the expected methods are implemented.

The only methods that appear to be exposed are the `get` and a `forEach` iterator.

> having to provide a manufacturer ID is quite awkward for a library user

Perhaps, but the company identifiers are ratified in the BLE specification so are not difficult to find. Both the JS and Android framework designers also appear to agree this is their preferred approach. Apple disagrees but they often do lol. ¯\\_(ツ)_/¯ 

The JS code to obtain the `keys` would use the iterator to traverse the entries, very straightforward in native JS, somewhat convoluted from Kotlin directly.

Comment (twyatt) [reply core/src/jsMain/kotlin/external/BluetoothAdvertisingEvent.kt:14]: Bummer it's not a straightforward / easy cast to `Map`.
Rather than using `Any?` (and then later having to use `asDynamic`) could we make this `BluetoothManufacturerDataMap` type?

Then define `BluetoothManufacturerDataMap` as something similar to?:

```kotlin
internal interface BluetoothManufacturerDataMap {
    fun get(key: String): ByteArray
    fun forEach(action: (key: String, value: ByteArray) -> Unit): Unit
}
```

Comment (davidtaylor-juul) [reply core/src/jsMain/kotlin/external/BluetoothAdvertisingEvent.kt:14]: Found a more generic mechanism by hooking into the javascript Object iterator interface and bridging it as a Kotlin `Iterable`. This lets us access any of the javascript methods that return `Iterator` such as `keys()` and `entries()` and typecast them as `Iterable`. This looks to be a less brittle approach and doesn't require any custom types.

See https://github.com/JuulLabs/kable/pull/41/commits/f6995fd5972c9931ba52079734ce8cad614e397e

File: core/src/jsMain/kotlin/Peripheral.kt [hunk 3]
```diff
@@ -250,4 +252,4 @@ public class JsPeripheral internal constructor(
     override fun toString(): String = "Peripheral(bluetoothDevice=${bluetoothDevice.string()})"
 }
 
-private fun ArrayBuffer.toByteArray(): ByteArray = Int8Array(this).unsafeCast<ByteArray>()
+internal fun ArrayBuffer.toByteArray(): ByteArray = Int8Array(this).unsafeCast<ByteArray>()
```
Comment (davidtaylor-juul) [inline core/src/jsMain/kotlin/Peripheral.kt:255]: "Borrowed" this extension for the advertisement data parsing. Should we move it somewhere else?

Comment (twyatt) [reply core/src/jsMain/kotlin/Peripheral.kt:253]: Yes please. Perhaps a new file names `Buffers.kt` or `Bytes.kt`?

File: core/src/jsMain/kotlin/Advertisement.kt [hunk 4]
```diff
@@ -1,12 +1,50 @@
...
+        get() = manufacturerDataEntries(advertisement.manufacturerData).firstOrNull()?.let { entry ->
+            ManufacturerData(
+                entry[0] as Short,
+                (entry[1] as DataView).buffer.toByteArray()
+            )
+        }
 }
+
+private fun manufacturerDataEntries(jsObject: dynamic): Iterable<Array<Any?>> =
+    (jsObject.entries() as JsIterator<Array<Any?>>).iterable()
```
Comment (twyatt) [inline core/src/jsMain/kotlin/Advertisement.kt:50]: We can remove the "unsafe" cast if we define the structure of the JS advertisement (in `BluetoothAdvertisingEvent.kt`):

```kotlin
internal interface BluetoothManufacturerDataMap {
    fun entries(): JsIterator<Array<Any?>>
}

/**
 * https://webbluetoothcg.github.io/web-bluetooth/#bluetoothadvertisingevent
 */
internal abstract external class BluetoothAdvertisingEvent : Event {
    val device: BluetoothDevice
    val uuids: Array<String>
    val name: String?
    val rssi: Int?
    val txPower: Int?
    val manufacturerData: BluetoothManufacturerDataMap
    val serviceData: Any?
}
```

```suggestion
    public actual val manufacturerData: ManufacturerData?
        get() = advertisement.manufacturerData.entries().iterable().firstOrNull()?.let { entry ->
            ManufacturerData(
                entry[0] as Short,
                (entry[1] as DataView).buffer.toByteArray()
            )
        }
}
```

Maybe we could also define the types of the entries so we don't have to do the casting when we create our `ManufacturerData`).

File: core/src/jsMain/kotlin/Advertisement.kt [hunk 5]
```diff
@@ -1,12 +1,50 @@
...
+    private val advertisement: BluetoothAdvertisingEvent,
 ) {
+    internal val bluetoothDevice: BluetoothDevice
+        get() = advertisement.device
 
     public actual val name: String?
         get() = bluetoothDevice.name
+
+    public actual val rssi: Int
+        get() = advertisement.rssi ?: 0
```
Comment (twyatt) [inline core/src/jsMain/kotlin/Advertisement.kt:20]: It's bizarre that JS has rssi as nullable. When would it ever not know RSSI for a received advertisement?! 🤷 

Not really liking falling back to `0`, as that is a valid rssi.
Perhaps we should fallback to `Interger.MIN_VALUE`)? Which is well outside the valid range (so could be conditionally checked for).

**Review by twyatt — COMMENTED** [review]
Re-request review after you've addressed comments and then I'll test on Android and macOS.

File: core/src/jsMain/kotlin/Advertisement.kt (hunk 5 above)
Comment (davidtaylor-juul) [reply core/src/jsMain/kotlin/Advertisement.kt:20]: Good catch. With a 32bit integer, using `Int.MIN_VALUE` as the RSSI a distance calculation would give a result of the order of 10^10^9 meters, which is essentially infinite. Or NaN if the overflow isn't dealt with 😄

**Comment by twyatt** [comment]
Looks great! I'll test the Android and macOS implementation today (then approve PR).
Aside from that, looks like you need to run `./gradlew apiDump`.

**Comment by twyatt** [comment]
Tested on Android:
_Log snippet from scan_

```
Forerunner 935 (E6:45:6C:C3:47:24)
  code=135
  data=0a 83
  name=Forerunner 935, rssi=-90
  Service UUID 6a4e3e10-667b-11e3-949a-0800200c9a66
    data=null
nxg200-30AEA4F48CE8 (30:AE:A4:F4:8C:EA)
  code=null
  data=null
  name=nxg200-30AEA4F48CE8, rssi=-87
  Service UUID 000000ff-0000-1000-8000-00805f9b34fb
    data=null
  Service UUID 000000ff-0000-1000-8000-00805f9b34fb
    data=null
N0404 (C9:F8:80:E1:2D:E6)
  code=null
  data=null
  name=N0404, rssi=-75
  Service UUID 0000feaf-0000-1000-8000-00805f9b34fb
    data=10 01 00 02 5a 23 09 00 b1 b6 c7 00 00 66 16 64 00
```

**Comment by twyatt** [comment]
macOS test:

```
Forerunner 935 (0BA3628E-1CF1-4E94-8A53-D1F1A4F696B6)
  code=null
  data=null
  name=Forerunner 935, rssi=-82
  Service UUID 6a4e3e10-667b-11e3-949a-0800200c9a66
    data=null

Forerunner 935 (0BA3628E-1CF1-4E94-8A53-D1F1A4F696B6)
  code=null
  data=null
  name=Forerunner 935, rssi=-79

null (D44A5F25-A784-4F00-A557-63BA894EEA21)
  code=null
  data=null
  name=null, rssi=-48
  Service UUID 0000fe9f-0000-1000-8000-00805f9b34fb
    data=00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```

**Comment by twyatt** [comment]
Possibly just different devices, but never say `code` or `data` populated from the macOS scan.

**Comment by twyatt** [comment]
Build is failing due to unused import, `./gradlew formatKotlin` should fix you up.

File: core/src/androidMain/kotlin/Advertisement.kt [hunk 6]
```diff
@@ -15,4 +20,27 @@ public actual class Advertisement(
 
     override fun toString(): String =
         "Advertisement(name=$name, rssi=$rssi, bluetoothDevice=$bluetoothDevice)"
```
[the remark this answers is not in the document]
Comment (twyatt) [reply core/src/androidMain/kotlin/Advertisement.kt:22]: Would also probably be a good idea to add `toString` overrides (for `Advertisement`) for Apple and JS targets.

**Comment by davidtaylor-juul** [comment]
> Possibly just different devices, but never saw `code` or `data` populated on the macOS scan

Possible, as Advertisment data can differ between packets even for the same device, but could be a bug. Investigating further and will add some tests.

**Comment by davidtaylor-juul** [comment]
> > Possibly just different devices, but never saw `code` or `data` populated on the macOS scan
> 
> Possible, as Advertisment data can differ between packets even for the same device, but could be a bug. Investigating further and will add some tests.

There was a bug, had `slice(2..size)` instead of `slice(2 until size)` which *should* have been throwing exceptions. Perhaps that test app is ignoring them?

Added some unit tests to cover that data parsing logic on macOS.

File: core/src/macosX64Main/kotlin/Advertisement.kt [hunk 7]
```diff
@@ -14,4 +21,39 @@ public actual data class Advertisement(
...
+        get() = (data[CBAdvertisementDataServiceUUIDsKey] as? List<CBUUID>)?.map { it.toUuid() } ?: emptyList()
+
+    public actual fun serviceData(uuid: Uuid): ByteArray? =
+        serviceDataAsNSData(uuid)?.toByteArray()
+
+    public fun serviceDataAsNSData(uuid: Uuid): NSData? =
+        (data[CBAdvertisementDataServiceDataKey] as? Map<CBUUID, NSData>)?.get(uuid.toCBUUID())
+
+    public actual fun manufacturerData(companyIdentifierCode: Int): ByteArray? =
+        manufacturerData?.takeIf { (it.code == companyIdentifierCode) }?.let { it.data }
```
Comment (twyatt) [inline core/src/macosX64Main/kotlin/Advertisement.kt:38]: Unnecessary `let`.

```suggestion
        manufacturerData?.takeIf { (it.code == companyIdentifierCode) }?.data
```

File: core/src/macosX64Main/kotlin/Advertisement.kt [hunk 8]
```diff
@@ -14,4 +21,39 @@ public actual data class Advertisement(
...
+
+    public actual val manufacturerData: ManufacturerData?
+        get() = manufacturerDataAsNSData?.toByteArray()?.toManufacturerData()
+
+    public fun manufacturerDataAsNSData(companyIdentifierCode: Int): NSData? =
+        manufacturerData(companyIdentifierCode)?.toNSData()
+
+    public val manufacturerDataAsNSData: NSData?
+        get() = data[CBAdvertisementDataServiceDataKey] as? NSData
```
Comment (twyatt) [inline core/src/macosX64Main/kotlin/Advertisement.kt:47]: Wrong key. This is why we were seeing `null` for `code` and `data`.

```suggestion
        get() = data[CBAdvertisementDataManufacturerDataKey] as? NSData
```

Thankfully the tests led me to investigate what data we were actually getting back, and voila! 😄

**Review by twyatt — COMMENTED** [review]
Amazing work! 💯 

Mostly nitpicks left (which you can optionally address) and [fix macOS key used for manufacturer data](https://github.com/JuulLabs/kable/pull/41#discussion_r549913043). Tested with correct key and it worked as expected (got back `code` and `data`).

File: core/src/macosX64Test/kotlin/AdvertisementTest.kt [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    @Test
+    fun manufacturerDataParsingWithFewerThan2BytesIsNull() {
+        val advertisement = fakeAdvertisement(
+            ubyteArrayOf(0x01u)
+        )
+        val data = advertisement.manufacturerData
+        assertNull(data)
+    }
+
+    private fun fakeAdvertisement(manufacturerBytes: UByteArray): Advertisement =
```
[the remark this answers is not in the document]
Comment (davidtaylor-juul) [reply core/src/macosX64Test/kotlin/AdvertisementTest.kt:51]: 👍 was super messy to specify the data using signed bytes. The lack of unsigned is part of *The Java Hangover* and we should not shy away from it for multi-platform targets IMHO. 😄

File: core/src/macosX64Main/kotlin/Advertisement.kt (hunk 8 above)
Comment (davidtaylor-juul) [reply core/src/macosX64Main/kotlin/Advertisement.kt:47]: OMG thanks for spotting this. I remember making sure to use the correct key originally and must have accidentally messed it up at some point 😭 Apple's bag-of-things API for advertisements is so vexing.

Comment (davidtaylor-juul) [reply core/src/macosX64Main/kotlin/Advertisement.kt:47]: I got it wrong in the tests also, at least my stupidity is consistent! 😑

File: core/src/macosX64Main/kotlin/Advertisement.kt (hunk 7 above)
Comment (davidtaylor-juul) [reply core/src/macosX64Main/kotlin/Advertisement.kt]: Good catch

**Review by twyatt — APPROVED** [review]
💯 🚀

File: core/api/core.api [hunk 10]
```diff
@@ -1,8 +1,13 @@
...
-	public fun <init> (ILandroid/bluetooth/BluetoothDevice;)V
+	public fun <init> (Landroid/bluetooth/le/ScanResult;)V
 	public final fun getAddress ()Ljava/lang/String;
+	public final fun getManufacturerData ()Lcom/juul/kable/ManufacturerData;
 	public final fun getName ()Ljava/lang/String;
 	public final fun getRssi ()I
+	public final fun getTxPower ()Ljava/lang/Integer;
+	public final fun getUuids ()Ljava/util/List;
+	public final fun manufacturerData (I)[B
+	public final fun serviceData (Ljava/util/UUID;)[B
```
Comment (Phoenix7351) [inline core/api/core.api:10]: For my own curiosity, what are these characters after the function definition? "(Ljava/util/UUID;)[B"

Something to do with transpilation, I assume, but what do they mean?

File: core/src/androidMain/kotlin/Advertisement.kt [hunk 11]
```diff
@@ -1,18 +1,46 @@
...
 
 import android.bluetooth.BluetoothDevice
+import android.bluetooth.le.ScanResult
+import android.os.ParcelUuid
+import com.benasher44.uuid.Uuid
 
 public actual class Advertisement(
-    public actual val rssi: Int,
-    internal val bluetoothDevice: BluetoothDevice,
+    private val scanResult: ScanResult,
```
Comment (Phoenix7351) [inline core/src/androidMain/kotlin/Advertisement.kt:9]: Nice.

File: core/src/jsMain/kotlin/Bytes.kt [hunk 12]
```diff
@@ -0,0 +1,6 @@
+package com.juul.kable
+
+import org.khronos.webgl.ArrayBuffer
+import org.khronos.webgl.Int8Array
+
+internal fun ArrayBuffer.toByteArray(): ByteArray = Int8Array(this).unsafeCast<ByteArray>()
```
Comment (Phoenix7351) [inline core/src/jsMain/kotlin/Bytes.kt:6]: Also nice.

File: core/src/jsMain/kotlin/external/JsIterator.kt [hunk 13]
```diff
@@ -0,0 +1,25 @@
+package com.juul.kable.external
+
+internal external interface JsIterator<T> {
+    fun next(): JsIteratorResult<T>
+}
+
+internal external interface JsIteratorResult<T> {
```
Comment (Phoenix7351) [inline core/src/jsMain/kotlin/external/JsIterator.kt:7]: "internal external" Haha. I get it, but funny.

Comment (Phoenix7351) [reply core/src/jsMain/kotlin/external/JsIterator.kt:7]: Question on this, though: "external" means, "the implementation is defined outside of kotlin," right?

If that's the case, how're we indicating to kotlin where that definition is?

File: core/src/jsMain/kotlin/external/JsIterator.kt [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    val done: Boolean
+    val value: T?
+}
+
+internal fun <T> JsIterator<T>.iterable(): Iterable<T> {
```
Comment (Phoenix7351) [inline core/src/jsMain/kotlin/external/JsIterator.kt:12]: Sucks that we need to manually define this.

File: core/src/macosX64Test/kotlin/AdvertisementTest.kt [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+import platform.CoreBluetooth.CBPeripheral
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertNotNull
+import kotlin.test.assertNull
+import kotlin.test.assertTrue
+
+class AdvertisementTest {
+    @Test
+    fun manufacturerDataCanContainData() {
```
Comment (Phoenix7351) [inline core/src/macosX64Test/kotlin/AdvertisementTest.kt:15]: For ease of understandability, I like to give my test function names a particular format:

fun [name of item being tested]_[input conditions]_[expected results]

I think in this case, it would be something like:

fun manufacturerData_validAdvertisement_returnsValidData()

Not saying that we need to use that exact format, but doing something akin to that makes it really easy to quickly understand what is being tested, under what conditions, and what the expected results are--a very useful thing to understand should the test ever fail.

File: core/api/core.api (hunk 10 above)
Comment (twyatt) [reply core/api/core.api:10]: Shorthand notation for Java bytecode.

[Here](http://homepage.divms.uiowa.edu/~slonnegr/wpj/JNI.pdf) is a short reference (for some of the symbols):
Screen Shot 2021-01-04 at 10 47 54 AM

Comment (twyatt) [reply core/api/core.api:10]: > Something to do with transpilation

Nothing to do w/ transpilation actually. Transpilation is for translating Kotlin to JavaScript.
There are Java (JVM) bytecode API dumps (not JavaScript at all).

File: core/src/jsMain/kotlin/external/JsIterator.kt (hunk 13 above)
Comment (twyatt) [reply core/src/jsMain/kotlin/external/JsIterator.kt:7]: > If that's the case, how're we indicating to kotlin where that definition is?

We actually don't have to. Since the Kotlin is transpiled to JavaScript, it will be running in the actually JavaScript environment (in the browser in our case). So we're simply promising the compiler that this function signature exists, and for us, it will exist when everything is transpiled. So the transpiled code will call this `interface` and as long as it actually matches the signature that does in fact exist, then all is good.

File: core/src/jsMain/kotlin/external/JsIterator.kt (hunk 14 above)
Comment (twyatt) [reply core/src/jsMain/kotlin/external/JsIterator.kt:12]: My hope is the new IR compiler (that JetBrains is working on) will provide this for us. We shall see though.

File: core/api/core.api (hunk 10 above)
Comment (Phoenix7351) [reply core/api/core.api:10]: Oh, why is this needed? The compiler can't figure this out for itself?

File: core/src/macosX64Test/kotlin/AdvertisementTest.kt (hunk 15 above)
Comment (twyatt) [reply core/src/macosX64Test/kotlin/AdvertisementTest.kt:15]: I kind of go back and forth on what the best format for test function names are.

> `[name of item being tested][input conditions][expected results]`

I used to subscribe the this format as well, but found it too restrictive at times.

I'm happy to keep this flexible (and let the developer decide what the best way to convey what the test is doing). As seeing that `manufacturerDataCanContainData` failed, seems reasonable enough to me to know what went wrong.

Happy if it is changed as well though. (I personally am fine with either naming scheme)

Comment (Phoenix7351) [reply core/src/macosX64Test/kotlin/AdvertisementTest.kt:15]: Well, what I find useful about using the "_", is that you're able to easily parse out all the components of the test (system under test, input conditions, expected results).

Camelcasing this info together will still contain these components, but it's more difficult to determine what they are, and could be interpreted different ways. 

In what way have you found the "_" format restrictive?

File: core/src/jsMain/kotlin/external/JsIterator.kt (hunk 13 above)
Comment (davidtaylor-juul) [reply core/src/jsMain/kotlin/external/JsIterator.kt:7]: FYI this protocol is part of the ECMAScript 2015 which is why we can define it as `external` and fully expect it to exist in the target runtime (for conforming browsers).

https://docs.w3cub.com/javascript/iteration_protocols#The_iterator_protocol

File: core/api/core.api (hunk 10 above)
Comment (twyatt) [reply core/api/core.api:10]: These API dumps (`.api`) are for our reference (not the compiler's). So we can observe how the API has changed. Simplifies the process of adhering to semver for our releases.

File: core/src/macosX64Test/kotlin/AdvertisementTest.kt (hunk 15 above)
Comment (twyatt) [reply core/src/macosX64Test/kotlin/AdvertisementTest.kt:15]: > In what way have you found the "_" format restrictive?

Some very small/simplified tests don't adhere to the structure:

1. name of item being tested
2. input conditions
3. expected results

Comment (davidtaylor-juul) [reply core/src/macosX64Test/kotlin/AdvertisementTest.kt:15]: As these are the first tests in this codebase I'll go ahead and switch to the `fun [name of item being tested]_[input conditions]_[expected results]` naming scheme to set precedent.

And thus defer further argument to our future selves. 🤠

Comment (Phoenix7351) [reply core/src/macosX64Test/kotlin/AdvertisementTest.kt]: > > In what way have you found the "_" format restrictive?
> 
> Some very small/simplified tests don't adhere to the structure:
> 
> 1. name of item being tested
> 2. input conditions
> 3. expected results

That's true. This very example could fit that case. Like, in, "manufacturerData_validAdvertisement_returnsValidData()", "validAdvertisement" might be superfluous (if there is a use case where there could be an invalid advertisement, I'd say it isn't). In that scenario, I think it'd be fine to shorten to `fun manufacturerData_returnsValidData()` as this still clearly denotes what is being tested and the outcome--with the input being understood as "happy path default conditions."

File: core/api/core.api (hunk 10 above)
Comment (Phoenix7351) [reply core/api/core.api:10]: Gotcha. Thanks!

File: core/src/macosX64Test/kotlin/AdvertisementTest.kt [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+            )
+        )
+        val data = advertisement.manufacturerData
+        assertNotNull(data)
+        assertEquals(data.code, 0x05c3)
+        assertTrue(data.data.isEmpty())
+    }
+
+    @Test
+    fun manufacturerData_advertisementWithFewerThanTwoBytes_isNull() {
```
Comment (Phoenix7351) [inline core/src/macosX64Test/kotlin/AdvertisementTest.kt:45]: https://www.youtube.com/watch?v=pso4SWM75ew

**Review by Phoenix7351 — APPROVED** [review]

File: core/src/macosX64Test/kotlin/AdvertisementTest.kt (hunk 15 above)
Comment (twyatt) [reply core/src/macosX64Test/kotlin/AdvertisementTest.kt]: Good example. Ultimately, I'm pretty flexible on the naming of test functions. 😄
